### DISCUSSION TITLE
CAN Sniffer Sensor - Add to Tasmota

### DISCUSSION THREAD

**Discussion by kwiatek6324** [arendst/Tasmota#14114] [Ideas]
Issue is a form of continuation for #7656

For my purpose i've created SENSOR based od mcp2515 and based on file xsns_87_mcp2515.ino
which for every incoming CAN packet will generate MQTT message, for example:

`/tasmota/stat/tasmota_xxxxxx/CAN {"ID":69,"LEN":5,"D0":255,"D1":71,"D2":0,"D3":0,"D4":119}`

Sending via mqtt with command: 

`/tasmota/cmnd/tasmota_xxxxxx/CanSend"  '{ "Id": 12, "Len": 8, "DATA": [ 111,21,1,0 ] }`

Sending from mqtt to can will be also possible (hope in 1-2 days).

My questions are:

1. How to add this feature to official tasmota repo ?
2. If by pull request, how to get ID for XSNS (for my purposes , i took 120)

Regards

**Comment by Jason2866** [comment]
yes, by Pull Request. XSNS 120 is okay. Theo will "adjust" to the actual next useable when he merges

**Comment by kwiatek6324** [comment]
Ok, I'll test it for next few days, then pull request.

Regards

**Reply by pkkrusty** [reply]
Did you ever move forward on this?

**Reply by pkkrusty** [reply]
@gemu2015 any interest in adding general canbus functionality to scripting based on mcp2515? Current driver is pretty limited.

**Reply by kwiatek6324** [reply]
Hi
@pkkrusty a little bit forget about this,
in 2 weeks i will make pull request.

I'm still using this on my HA installation
since dec 2021.

Regards

**Reply by pkkrusty** [reply]
Awesome, thanks. That will greatly expand usefulness of the current driver.

**Reply by pkkrusty** [reply]
Making progress?

**Reply by barbudor** [reply]
Just wondering what are your use case for CAN in Tasmota
I would assume that CAN is used in car where you don't necesseraly have Wifi available
Is that used in other cases ?
Thanks

**Reply by pkkrusty** [reply]
Communicating with BMS on a electric car battery. VW Type 5. They have Canbus and balancing built-in to each module.

**Reply by barbudor** [reply]
Ok, Tasmota connect to the home wifi but only when car is at home ?

**Reply by pkkrusty** [reply]
Car battery is removed from car. Will be used for small solar installation.

**Reply by barbudor** [reply]
oh yes, I didn't linked to your other post 😄

**Reply by ortegafernando** [reply]
In a car, this "every incoming CAN packet will generate MQTT message" will generate huge messages, every several miliseconds we are receiving CAN packets, so may be you can implement filters (include or exclude) before generating MQTT message.

**Reply by pkkrusty** [reply]
> Hi @pkkrusty a little bit forget about this, in 2 weeks i will make pull request.
> 
> I'm still using this on my HA installation since dec 2021.
> 
> Regards

@kwiatek6324 Hate to keep bugging you, but excited to try out whatever you have put together.

**Reply by pkkrusty** [reply]
@kwiatek6324 Still interested in this!

**Comment by goodkiller** [comment]
Related with #18287
